src/applications/disability-benefits/526EZ/components/AsyncDisplayWidget.jsx

```diff
@@ -0,0 +1,104 @@
+import React from 'react';
+
+import LoadingIndicator from '@department-of-veterans-affairs/formation/LoadingIndicator';
+import AlertBox from '@department-of-veterans-affairs/formation/AlertBox';
+
+const PENDING = 'PENDING';
+const RESOLVED = 'RESOLVED';
+const REJECTED = 'REJECTED';
+
+
+/**
+ * Handles the various display statuses when calling an asynchronous function.
+ */
+export default class AsyncDisplayWidget extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
+      data: null,
+      promiseState: PENDING
+    };
+
+    // Make sure we configured it correctly
+    // StringField passes 'ui:options' as the options prop
+    const uiOptions = props.options;
+    if (!uiOptions) {
+      throw new Error('No ui:options supplied to AsyncDisplayWidget.');
+    }
+
+    if (typeof uiOptions.viewComponent !== 'function') {
+      throw new Error('AsyncDisplayWidget requires viewComponent in ui:options to be a React component.');
+    }
+
+    if (typeof uiOptions.callback !== 'function') {
+      throw new Error('AsyncDisplayWidget requires callback in ui:options to be a function.');
+    }
+
+    // TODO: Don't call the callback _every_ time the component is mounted
+    const cbPromise = uiOptions.callback();
+    if (cbPromise instanceof Promise) {
+      cbPromise.then((data) => {
```

Comment:
Honestly not sure what the implications of calling `setState` within a promise within the constructor are, but I do know that React generally discourages use of `setState` and side effects within the constructor. Might it make sense to check that the cb is a promise in the constructor, but offload the actual call to `componentDidMount`?

Reply:
Well, we can't check that the callback returns a promise without calling it, so it's either all in the constructor or all in `componentDidMount`. That said, I think `componentDidMount` is a great place to put this. :+1: